File: JumpyOtter/Sources/Terrain.swift
import SceneKit
import UIKit

enum RowKind {
    case grass, road, river, rail
}

enum TrainPhase {
    case idle, warning, running
}

final class MovingObject {
    let node: SCNNode
    var x: Float
    let halfLen: Float
    var presentationX: Float { node.presentation.position.x }
    init(node: SCNNode, x: Float, halfLen: Float) {
        self.node = node
        self.x = x
        self.halfLen = halfLen
    }
}

final class Row {
    let index: Int
    let kind: RowKind
    let node = SCNNode()

    var dir: Float = 1
    var speed: Float = 0
    var objects: [MovingObject] = []
    var blocked: Set<Int> = []
    var creatine: [Int: SCNNode] = [:]

    // rail state
    var trainPhase: TrainPhase = .idle
    var trainTimer: Float = 0
    var train: MovingObject?
    var signal: SCNNode?
    var blinkTimer: Float = 0
    var blinkOn = false

    init(index: Int, kind: RowKind) {
        self.index = index
        self.kind = kind
        node.position = SCNVector3(0, 0, Float(index))
    }

    /// Surface height the player stands on for this row.
    var surfaceY: Float { kind == .river ? 0.11 : 0 }
}

/// Generates rows procedurally and builds their scenery.
final class TerrainGenerator {

    private(set) var rows: [Int: Row] = [:]
    private let parent: SCNNode
    private var nextIndex = 0
    private var minIndex = 0  // lowest row index ever created

    // group state so we emit runs of the same kind (multi-lane roads etc.)
    private var pendingKind: RowKind = .grass
    private var pendingCount = 0
    private var lastGroupKind: RowKind = .grass
    private var lastRiverDir: Float = 1

    init(parent: SCNNode) {
        self.parent = parent
    }

    func reset() {
        for (_, row) in rows { row.node.removeFromParentNode() }
        rows.removeAll()
        nextIndex = 0
        minIndex = 0
        pendingKind = .grass
        pendingCount = 0
        lastGroupKind = .grass
    }

    /// Ensure rows exist for indices `from...to`; cull rows below `from`.
    func ensure(from: Int, to: Int) {
        for (idx, row) in rows where idx < from {
            row.node.removeFromParentNode()
            rows.removeValue(forKey: idx)
        }
        // backfill grass forest behind the start (runs once; culled rows are never re-created
        // because minIndex tracks the lowest index ever built)
        while minIndex > from {
            minIndex -= 1
            let row = buildRow(index: minIndex, kind: .grass)
            rows[minIndex] = row
            parent.addChildNode(row.node)
        }
        while nextIndex <= to {
            let row = makeRow(at: nextIndex)
            rows[nextIndex] = row
            parent.addChildNode(row.node)
            nextIndex += 1
        }
    }

    // MARK: - Row creation

    private func makeRow(at index: Int) -> Row {
        if index <= 0 { return buildRow(index: index, kind: .grass) }
        if pendingCount == 0 { rollNextGroup(at: index) }
        pendingCount -= 1
        return buildRow(index: index, kind: pendingKind)
    }

    private func rollNextGroup(at index: Int) {
        if lastGroupKind != .grass {
            pendingKind = .grass
            pendingCount = 1
            lastGroupKind = .grass
            return
        }
        // Difficulty: hazards become more frequent with distance.
        let difficulty = min(Float(index) / 150.0, 1.0)
        for _ in 0..<8 {
            let r = Float.random(in: 0..<1)
            let grassChance: Float = 0.34 - 0.12 * difficulty
            let roadChance: Float = 0.30 + 0.06 * difficulty
            let riverChance: Float = 0.22 + 0.04 * difficulty
            if r < grassChance {
                pendingKind = .grass
                pendingCount = Int.random(in: 1...2)
            } else if r < grassChance + roadChance {
                pendingKind = .road
                pendingCount = Int.random(in: 1...(index > 12 ? 3 : 2))
            } else if r < grassChance + roadChance + riverChance {
                pendingKind = .river
                pendingCount = Int.random(in: 1...(index > 20 ? 3 : 2))
                lastRiverDir = Bool.random() ? 1 : -1
            } else {
                pendingKind = .rail
                pendingCount = 1
            }
            // avoid back-to-back hazard groups of the same kind (mega-roads etc.)
            if pendingKind == .grass || pendingKind != lastGroupKind { break }
        }
        lastGroupKind = pendingKind
    }

    private func buildRow(index: Int, kind: RowKind) -> Row {
        let row = Row(index: index, kind: kind)
        switch kind {
        case .grass: buildGrass(row)
        case .road:  buildRoad(row)
        case .river: buildRiver(row)
        case .rail:  buildRail(row)
        }
        return row
    }

    // MARK: - Grass

    private func buildGrass(_ row: Row) {
        let light = row.index % 2 == 0
        let base = VoxelFactory.box(
            w: CGFloat(K.visualHalfWidth * 2), h: 0.5, l: 1.0,
            color: light ? Palette.grassLight : Palette.grassDark,
            y: -0.25, chamfer: 0)
        row.node.addChildNode(base)
        // darker side strips outside the playable area
        for sx in [Float(1), -1] {
            let stripW = K.visualHalfWidth - Float(K.maxCol) - 0.5
            let side = VoxelFactory.box(
                w: CGFloat(stripW), h: 0.04, l: 1.0,
                color: Palette.sidewalk,
                x: sx * (Float(K.maxCol) + 0.5 + stripW / 2),
                y: 0.02, chamfer: 0)
            side.castsShadow = false
            row.node.addChildNode(side)
            row.node.addChildNode(VoxelFactory.box(
                w: 0.12, h: 0.10, l: 1.0, color: Palette.curb,
                x: sx * (Float(K.maxCol) + 0.56), y: 0.05, chamfer: 0.01))
        }

        // colorful homes and trees outside the playable strip
        for col in stride(from: Int(-K.visualHalfWidth) + 1, through: Int(K.visualHalfWidth) - 1, by: 1) {
            if col >= K.minCol - 1 && col <= K.maxCol + 1 { continue }
            let roll = Float.random(in: 0..<1)
            if roll < 0.52 {
                let house = VoxelFactory.paintedLady(color: Palette.houseColors.randomElement()!)
                let scale = Float.rand(0.82...1.08)
                house.scale = SCNVector3(scale, scale, scale)
                house.position = SCNVector3(Float(col), 0, 0)
                row.node.addChildNode(house)
            } else if roll < 0.70 {
                let t = VoxelFactory.tree(height: Int.random(in: 2...3))
                t.position = SCNVector3(Float(col), 0, 0)
                row.node.addChildNode(t)
            }
        }

        // wildflowers and grass tufts inside the playable strip
        for col in K.minCol...K.maxCol where Float.random(in: 0..<1) < 0.28 {
            let ox = Float(col) + Float.rand(-0.32...0.32)
            let oz = Float.rand(-0.32...0.32)
            let deco: SCNNode
            if Float.random(in: 0..<1) < 0.55 {
                deco = VoxelFactory.flower(color: Palette.flowerColors.randomElement()!)
            } else {
                deco = VoxelFactory.grassTuft()
            }
            deco.position = SCNVector3(ox, 0, oz)
            deco.eulerAngles.y = Float.rand(0...Float.pi)
            row.node.addChildNode(deco)
        }

        // trees inside the playable strip
        guard row.index != 0 else { return }  // keep spawn row clear
        var cols = Array(K.minCol...K.maxCol).shuffled()
        let treeCount = row.index < 0 ? Int.random(in: 2...4) : Int.random(in: 0...3)
        var placed = 0
        for col in cols {
            if placed >= treeCount { break }
            // always keep column 0 clear on the first few rows so the player has a path
            if row.index > 0 && row.index <= 3 && col == 0 { continue }
            if row.blocked.count >= 6 { break }  // never wall off the row
            let t = VoxelFactory.tree(height: Int.random(in: 1...3))
            t.position = SCNVector3(Float(col), 0, 0)
            row.node.addChildNode(t)
            row.blocked.insert(col)
            placed += 1
        }
        // creatine chance
        if row.index > 0, Float.random(in: 0..<1) < 0.12 {
            cols = Array(K.minCol...K.maxCol).filter { !row.blocked.contains($0) }.shuffled()
            if let col = cols.first {
                let c = VoxelFactory.creatineBottle()
                c.position.x = Float(col)
                row.node.addChildNode(c)
                row.creatine[col] = c
            }
        }
    }

    // MARK: - Road

    private func buildRoad(_ row: Row) {
        let base = VoxelFactory.box(
            w: CGFloat(K.visualHalfWidth * 2), h: 0.5, l: 1.0,
            color: Palette.road, y: -0.25, chamfer: 0)
        row.node.addChildNode(base)
        for z in [Float(-0.18), 0.18] {
            row.node.addChildNode(VoxelFactory.box(
                w: CGFloat(K.visualHalfWidth * 2), h: 0.025, l: 0.035,
                color: Palette.railSteel, y: 0.013, z: z, chamfer: 0))
        }

        // lane divider dashes if the previous row was also road
        if rows[row.index - 1]?.kind == .road {
            var x: Float = -K.visualHalfWidth + 0.5
            while x < K.visualHalfWidth {
                let dash = VoxelFactory.box(w: 0.45, h: 0.02, l: 0.08,
                                            color: UIColor(white: 1, alpha: 0.85),
                                            x: x, y: 0.011, z: -0.5, chamfer: 0)
                row.node.addChildNode(dash)
                x += 1.4
            }
        }

        let difficulty = min(Float(max(row.index, 0)) / 120.0, 1.0)
        row.dir = Bool.random() ? 1 : -1
        row.speed = Float.rand(1.8...3.4) + difficulty * 1.6

        // spawn traffic spread around the wrap track
        var x: Float = -K.wrapHalf + Float.rand(0...2)
        while x < K.wrapHalf - 2 {
            let vehicleRoll = Float.random(in: 0..<1)
            let vehicle: (SCNNode, Float)
            if vehicleRoll < 0.20 {
                vehicle = VoxelFactory.cableCar()
            } else if vehicleRoll < 0.36 {
                vehicle = VoxelFactory.truck()
            } else {
                vehicle = VoxelFactory.car(color: Palette.carColors.randomElement()!)
            }
            let (node, halfLen) = vehicle
            if row.dir < 0 { node.eulerAngles.y = .pi }
            x += halfLen
            node.position = SCNVector3(x, 0, 0)
            row.node.addChildNode(node)
            row.objects.append(MovingObject(node: node, x: x, halfLen: halfLen))
            x += halfLen + Float.rand(3.4...9.0) - difficulty * 1.6
        }

        // creatine chance on road
        if row.index > 0, Float.random(in: 0..<1) < 0.08 {
            let col = Int.random(in: K.minCol...K.maxCol)
            let c = VoxelFactory.creatineBottle()
            c.position.x = Float(col)
            row.node.addChildNode(c)
            row.creatine[col] = c
        }
    }

    // MARK: - River

    private func buildRiver(_ row: Row) {
        // river bed
        let bed = VoxelFactory.box(
            w: CGFloat(K.visualHalfWidth * 2), h: 0.3, l: 1.0,
            color: Palette.riverBed, y: -0.45, chamfer: 0)
        row.node.addChildNode(bed)
        // water surface (slightly translucent)
        let water = VoxelFactory.box(
            w: CGFloat(K.visualHalfWidth * 2), h: 0.16, l: 1.0,
            color: Palette.water, y: -0.18, chamfer: 0)
        water.geometry?.firstMaterial?.transparency = 0.92
        water.castsShadow = false
        row.node.addChildNode(water)
        // foam streaks drifting with the current
        var fx: Float = -K.visualHalfWidth + Float.rand(0.2...1.4)
        while fx < K.visualHalfWidth - 0.6 {
            let len = CGFloat(Float.rand(0.35...0.9))
            let foam = VoxelFactory.box(w: len, h: 0.02, l: 0.06, color: Palette.waterFoam,
                                        x: fx, y: -0.095, z: Float.rand(-0.4...0.4), chamfer: 0)
            foam.castsShadow = false
            foam.opacity = 0.75
            let drift = CGFloat(Float.rand(0.8...1.6))
            foam.runAction(.repeatForever(.sequence([
                .moveBy(x: drift, y: 0, z: 0, duration: TimeInterval(Float.rand(1.6...2.8))),
                .fadeOut(duration: 0.3),
                .moveBy(x: -drift, y: 0, z: 0, duration: 0),
                .fadeOpacity(to: 0.75, duration: 0.3),
            ])))
            row.node.addChildNode(foam)
            fx += Float.rand(1.8...3.6)
        }

        // alternate direction within river groups
        lastRiverDir = -lastRiverDir
        row.dir = lastRiverDir
        row.speed = Float.rand(1.0...2.1)

        var x: Float = -K.wrapHalf + Float.rand(0...1.5)
        while x < K.wrapHalf - 2 {
            let tiles = Int.random(in: 2...4)
            let (node, halfLen) = VoxelFactory.ferry(tiles: tiles)
            x += halfLen
            node.position = SCNVector3(x, 0.0, 0)
            row.node.addChildNode(node)
            row.objects.append(MovingObject(node: node, x: x, halfLen: halfLen))
            x += halfLen + Float.rand(1.6...3.4)
        }
    }

    // MARK: - Rail

    private func buildRail(_ row: Row) {
        let base = VoxelFactory.box(
            w: CGFloat(K.visualHalfWidth * 2), h: 0.5, l: 1.0,
            color: Palette.railBed, y: -0.25, chamfer: 0)
        row.node.addChildNode(base)
        // ties
        var x: Float = -K.visualHalfWidth + 0.3
        while x < K.visualHalfWidth {
            row.node.addChildNode(VoxelFactory.box(w: 0.28, h: 0.06, l: 0.8,
                                                   color: Palette.railTie, x: x, y: 0.03, chamfer: 0))
            x += 0.62
        }
        // rails
        for dz in [Float(0.22), -0.22] {
            row.node.addChildNode(VoxelFactory.box(w: CGFloat(K.visualHalfWidth * 2), h: 0.07, l: 0.09,
                                                   color: Palette.railSteel, y: 0.085, z: dz, chamfer: 0))
        }
        // signal at the side of the playable area
        let signal = VoxelFactory.railSignal()
        signal.position = SCNVector3(Float(K.maxCol) + 1.4, 0, -0.2)
        row.node.addChildNode(signal)
        row.signal = signal

        row.dir = Bool.random() ? 1 : -1
        row.trainPhase = .idle
        row.trainTimer = Float.rand(1.5...5.0)
    }
}

File: android/app/src/main/java/com/devin/jumpyotter/Terrain.kt
package com.devin.jumpyotter

import com.devin.jumpyotter.engine.FadeTo
import com.devin.jumpyotter.engine.MeshBuilder
import com.devin.jumpyotter.engine.MoveBy
import com.devin.jumpyotter.engine.Node
import com.devin.jumpyotter.engine.RepeatForever
import com.devin.jumpyotter.engine.Sequence
import kotlin.math.PI

enum class RowKind { GRASS, ROAD, RIVER, RAIL }

enum class TrainPhase { IDLE, WARNING, RUNNING }

class MovingObject(val node: Node, var x: Float, val halfLen: Float) {
    val presentationX: Float get() = node.position.x
}

class Row(val index: Int, val kind: RowKind) {
    val node = Node()

    var dir = 1f
    var speed = 0f
    val objects = ArrayList<MovingObject>()
    val blocked = HashSet<Int>()
    val creatine = HashMap<Int, Node>()

    // rail state
    var trainPhase = TrainPhase.IDLE
    var trainTimer = 0f
    var train: MovingObject? = null
    var signal: Node? = null
    var blinkTimer = 0f
    var blinkOn = false

    init {
        node.position.set(0f, 0f, index.toFloat())
    }

    /** Surface height the player stands on for this row. */
    val surfaceY: Float get() = if (kind == RowKind.RIVER) 0.11f else 0f
}

/** Generates rows procedurally and builds their scenery. Static scenery per row is merged into one mesh. */
class TerrainGenerator(private val parent: Node, private val onRowRemoved: (Row) -> Unit) {

    val rows = HashMap<Int, Row>()
    private var nextIndex = 0
    private var minIndex = 0  // lowest row index ever created

    // group state so we emit runs of the same kind (multi-lane roads etc.)
    private var pendingKind = RowKind.GRASS
    private var pendingCount = 0
    private var lastGroupKind = RowKind.GRASS
    private var lastRiverDir = 1f

    fun reset() {
        for (row in rows.values) {
            row.node.removeFromParent()
            onRowRemoved(row)
        }
        rows.clear()
        nextIndex = 0
        minIndex = 0
        pendingKind = RowKind.GRASS
        pendingCount = 0
        lastGroupKind = RowKind.GRASS
    }

    /** Ensure rows exist for indices `from..to`; cull rows below `from`. */
    fun ensure(from: Int, to: Int) {
        val it = rows.entries.iterator()
        while (it.hasNext()) {
            val (idx, row) = it.next()
            if (idx < from) {
                row.node.removeFromParent()
                onRowRemoved(row)
                it.remove()
            }
        }
        // backfill grass forest behind the start (runs once; culled rows are never re-created
        // because minIndex tracks the lowest index ever built)
        while (minIndex > from) {
            minIndex -= 1
            val row = buildRow(minIndex, RowKind.GRASS)
            rows[minIndex] = row
            parent.addChild(row.node)
        }
        while (nextIndex <= to) {
            val row = makeRow(nextIndex)
            rows[nextIndex] = row
            parent.addChild(row.node)
            nextIndex += 1
        }
    }

    // MARK: - Row creation

    private fun makeRow(index: Int): Row {
        if (index <= 0) return buildRow(index, RowKind.GRASS)
        if (pendingCount == 0) rollNextGroup(index)
        pendingCount -= 1
        return buildRow(index, pendingKind)
    }

    private fun rollNextGroup(index: Int) {
        if (lastGroupKind != RowKind.GRASS) {
            pendingKind = RowKind.GRASS
            pendingCount = 1
            lastGroupKind = RowKind.GRASS
            return
        }
        // Difficulty: hazards become more frequent with distance.
        val difficulty = minOf(index / 150f, 1f)
        for (attempt in 0 until 8) {
            val r = rand01()
            val grassChance = 0.34f - 0.12f * difficulty
            val roadChance = 0.30f + 0.06f * difficulty
            val riverChance = 0.22f + 0.04f * difficulty
            if (r < grassChance) {
                pendingKind = RowKind.GRASS
                pendingCount = randInt(1, 2)
            } else if (r < grassChance + roadChance) {
                pendingKind = RowKind.ROAD
                pendingCount = randInt(1, if (index > 12) 3 else 2)
            } else if (r < grassChance + roadChance + riverChance) {
                pendingKind = RowKind.RIVER
                pendingCount = randInt(1, if (index > 20) 3 else 2)
                lastRiverDir = if (randBool()) 1f else -1f
            } else {
                pendingKind = RowKind.RAIL
                pendingCount = 1
            }
            // avoid back-to-back hazard groups of the same kind (mega-roads etc.)
            if (pendingKind == RowKind.GRASS || pendingKind != lastGroupKind) break
        }
        lastGroupKind = pendingKind
    }

    private fun buildRow(index: Int, kind: RowKind): Row {
        val row = Row(index, kind)
        val mb = MeshBuilder()
        when (kind) {
            RowKind.GRASS -> buildGrass(row, mb)
            RowKind.ROAD -> buildRoad(row, mb)
            RowKind.RIVER -> buildRiver(row, mb)
            RowKind.RAIL -> buildRail(row, mb)
        }
        if (!mb.isEmpty) row.node.addChild(Node(mb.build()))
        return row
    }

    // MARK: - Grass

    private fun buildGrass(row: Row, mb: MeshBuilder) {
        val light = row.index % 2 == 0
        mb.box(K.visualHalfWidth * 2, 0.5f, 1.0f, if (light) Palette.grassLight else Palette.grassDark, 0f, -0.25f, 0f)
        // darker side strips outside the playable area
        for (sx in floatArrayOf(1f, -1f)) {
            val stripW = K.visualHalfWidth - K.maxCol - 0.5f
            mb.box(stripW, 0.04f, 1.0f, Palette.sidewalk, sx * (K.maxCol + 0.5f + stripW / 2), 0.02f, 0f)
            mb.box(0.12f, 0.10f, 1.0f, Palette.curb, sx * (K.maxCol + 0.56f), 0.05f, 0f)
        }

        // colorful homes and trees outside the playable strip
        for (col in (-K.visualHalfWidth.toInt() + 1)..(K.visualHalfWidth.toInt() - 1)) {
            if (col >= K.minCol - 1 && col <= K.maxCol + 1) continue
            val roll = rand01()
            if (roll < 0.52f) {
                val scale = rand(0.82f, 1.08f)
                mb.withOffset(col.toFloat(), 0f, 0f, scale) {
                    VoxelFactory.paintedLadyInto(this, Palette.houseColors.random())
                }
            } else if (roll < 0.70f) {
                mb.withOffset(col.toFloat(), 0f, 0f) { VoxelFactory.treeInto(this, randInt(2, 3)) }
            }
        }

        // wildflowers and grass tufts inside the playable strip
        for (col in K.minCol..K.maxCol) {
            if (rand01() >= 0.28f) continue
            val ox = col + rand(-0.32f, 0.32f)
            val oz = rand(-0.32f, 0.32f)
            mb.withOffset(ox, 0f, oz, 1f, rand(0f, PI.toFloat())) {
                if (rand01() < 0.55f) VoxelFactory.flowerInto(this, Palette.flowerColors.random())
                else VoxelFactory.grassTuftInto(this)
            }
        }

        // trees inside the playable strip
        if (row.index == 0) return  // keep spawn row clear
        var cols = (K.minCol..K.maxCol).shuffled()
        val treeCount = if (row.index < 0) randInt(2, 4) else randInt(0, 3)
        var placed = 0
        for (col in cols) {
            if (placed >= treeCount) break
            // always keep column 0 clear on the first few rows so the player has a path
            if (row.index > 0 && row.index <= 3 && col == 0) continue
            if (row.blocked.size >= 6) break  // never wall off the row
            mb.withOffset(col.toFloat(), 0f, 0f) { VoxelFactory.treeInto(this, randInt(1, 3)) }
            row.blocked.add(col)
            placed += 1
        }
        // creatine chance
        if (row.index > 0 && rand01() < 0.12f) {
            cols = (K.minCol..K.maxCol).filter { !row.blocked.contains(it) }.shuffled()
            cols.firstOrNull()?.let { col ->
                val c = VoxelFactory.creatineBottle()
                c.position.x = col.toFloat()
                row.node.addChild(c)
                row.creatine[col] = c
            }
        }
    }

    // MARK: - Road

    private fun buildRoad(row: Row, mb: MeshBuilder) {
        mb.box(K.visualHalfWidth * 2, 0.5f, 1.0f, Palette.road, 0f, -0.25f, 0f)
        for (z in floatArrayOf(-0.18f, 0.18f)) {
            mb.box(K.visualHalfWidth * 2, 0.025f, 0.035f, Palette.railSteel, 0f, 0.013f, z)
        }

        // lane divider dashes if the previous row was also road
        if (rows[row.index - 1]?.kind == RowKind.ROAD) {
            var x = -K.visualHalfWidth + 0.5f
            while (x < K.visualHalfWidth) {
                mb.box(0.45f, 0.02f, 0.08f, Rgb.gray(0.95f), x, 0.011f, -0.5f)
                x += 1.4f
            }
        }

        val difficulty = minOf(maxOf(row.index, 0) / 120f, 1f)
        row.dir = if (randBool()) 1f else -1f
        row.speed = rand(1.8f, 3.4f) + difficulty * 1.6f

        // spawn traffic spread around the wrap track
        var x = -K.wrapHalf + rand(0f, 2f)
        while (x < K.wrapHalf - 2) {
            val vehicleRoll = rand01()
            val vehicle = when {
                vehicleRoll < 0.20f -> VoxelFactory.cableCar()
                vehicleRoll < 0.36f -> VoxelFactory.truck()
                else -> VoxelFactory.car(Palette.carColors.random())
            }
            if (row.dir < 0) vehicle.node.eulerAngles.y = PI.toFloat()
            x += vehicle.halfLen
            vehicle.node.position.set(x, 0f, 0f)
            row.node.addChild(vehicle.node)
            row.objects.add(MovingObject(vehicle.node, x, vehicle.halfLen))
            x += vehicle.halfLen + rand(3.4f, 9.0f) - difficulty * 1.6f
        }

        // creatine chance on road
        if (row.index > 0 && rand01() < 0.08f) {
            val col = randInt(K.minCol, K.maxCol)
            val c = VoxelFactory.creatineBottle()
            c.position.x = col.toFloat()
            row.node.addChild(c)
            row.creatine[col] = c
        }
    }

    // MARK: - River

    private fun buildRiver(row: Row, mb: MeshBuilder) {
        // river bed and water surface
        mb.box(K.visualHalfWidth * 2, 0.3f, 1.0f, Palette.riverBed, 0f, -0.45f, 0f)
        mb.box(K.visualHalfWidth * 2, 0.16f, 1.0f, Palette.water, 0f, -0.18f, 0f)
        // foam streaks drifting with the current
        var fx = -K.visualHalfWidth + rand(0.2f, 1.4f)
        while (fx < K.visualHalfWidth - 0.6f) {
            val foam = VoxelFactory.box(rand(0.35f, 0.9f), 0.02f, 0.06f, Palette.waterFoam, fx, -0.095f, rand(-0.4f, 0.4f))
            foam.opacity = 0.75f
            val drift = rand(0.8f, 1.6f)
            foam.runAction(
                RepeatForever(
                    Sequence(
                        MoveBy(drift, 0f, 0f, rand(1.6f, 2.8f)),
                        FadeTo(0f, 0.3f),
                        MoveBy(-drift, 0f, 0f, 0f),
                        FadeTo(0.75f, 0.3f),
                    )
                )
            )
            row.node.addChild(foam)
            fx += rand(1.8f, 3.6f)
        }

        // alternate direction within river groups
        lastRiverDir = -lastRiverDir
        row.dir = lastRiverDir
        row.speed = rand(1.0f, 2.1f)

        var x = -K.wrapHalf + rand(0f, 1.5f)
        while (x < K.wrapHalf - 2) {
            val tiles = randInt(2, 4)
            val ferry = VoxelFactory.ferry(tiles)
            x += ferry.halfLen
            ferry.node.position.set(x, 0f, 0f)
            row.node.addChild(ferry.node)
            row.objects.add(MovingObject(ferry.node, x, ferry.halfLen))
            x += ferry.halfLen + rand(1.6f, 3.4f)
        }
    }

    // MARK: - Rail

    private fun buildRail(row: Row, mb: MeshBuilder) {
        mb.box(K.visualHalfWidth * 2, 0.5f, 1.0f, Palette.railBed, 0f, -0.25f, 0f)
        // ties
        var x = -K.visualHalfWidth + 0.3f
        while (x < K.visualHalfWidth) {
            mb.box(0.28f, 0.06f, 0.8f, Palette.railTie, x, 0.03f, 0f)
            x += 0.62f
        }
        // rails
        for (dz in floatArrayOf(0.22f, -0.22f)) {
            mb.box(K.visualHalfWidth * 2, 0.07f, 0.09f, Palette.railSteel, 0f, 0.085f, dz)
        }
        // signal at the side of the playable area
        val signal = VoxelFactory.railSignal()
        signal.position.set(K.maxCol + 1.4f, 0f, -0.2f)
        row.node.addChild(signal)
        row.signal = signal

        row.dir = if (randBool()) 1f else -1f
        row.trainPhase = TrainPhase.IDLE
        row.trainTimer = rand(1.5f, 5.0f)
    }
}
